src/Provider/AbstractProvider.php

```diff
@@ -686,18 +702,26 @@ protected function parseResponse(ResponseInterface $response)
             return $this->parseJson($content);
         } catch (UnexpectedValueException $e) {
             if (strpos($type, 'json') !== false) {
-                throw $e;
+                throw $this->mayThrowResponseParsingException
+                    ? new ResponseParsingException($response, $content, $e->getMessage(), $e->getCode())
+                    : $e;
             }
 
-            if ($response->getStatusCode() == 500) {
-                throw new UnexpectedValueException(
-                    'An OAuth server error was encountered that did not contain a JSON body',
-                    0,
-                    $e
-                );
+            // for any other content types
+            if ($this->mayThrowResponseParsingException) {
```

Comment:
Maybe we always throw the exception, no matter what?

I know this suggestion is somewhat of a BC break, but if you have `ResponseParsingException` extend `UnexpectedValueException`, then anyone catching that exception will start catching the new exceptions and can decide what to do about them.

@shadowhand, @stevenmaguire, thoughts?

Reply:
I like your suggestion. I can implement the changes if no objection coming from @shadowhand and @stevenmaguire